File: lvl-3a--fuel_injection_perfection/solution_3a.py
import pdb



import math
import sys
sys.setrecursionlimit(1500)

def answer(n):
  n = int(n)
  return int(find_steps_count(n, 0)) # don't forget to cast to int. Type matters!

def base_cases(num):
  if num == 3:
    return 2
  elif num == 2:
    return 1
  elif num <= 1:
    return 0
  else:
    return None

def find_steps_count(num, current_count = 0):
  base_result = base_cases(num)
  if base_result != None:
    return current_count + base_result

  current_count = current_count + 1
  if num % 2 == 0:
    return find_steps_count(num / 2, current_count)
  else:
    new_num = (num + 1) if second_last_digit_is_one(num) else (num - 1)
    return find_steps_count(new_num, current_count)

# Check if second last digit in binary is 1
def second_last_digit_is_one(num):
  return (num >> 1) % 2 != 0



result = answer( sys.argv[1] )
print 'RESULT: ', result

File: lvl-3a--fuel_injection_perfection/solution_3a.rb
class Solution_3a

  def answer(num_str)
    num = num_str.to_i
    find_steps_count(num, 0)
  end

  def base_cases(num)
    if num == 3
      2
    elsif num == 2
      1
    elsif num == 1
      0
    else
      nil
    end
  end

  def find_steps_count(num, current_count = 0)
    if base_result = base_cases(num)
      return current_count + base_result
    end

    if num.even?
      return current_count + 1 + find_steps_count(num / 2, current_count)
    else
      new_num = second_last_digit_is_one?(num) ? (num + 1) : (num - 1)
      return current_count + 1 + find_steps_count(new_num, current_count)
    end
  end

  def second_last_digit_is_one?(num)
    (num >> 1).odd?
  end
end
